{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfd25534",
   "metadata": {},
   "outputs": [],
   "source": [
    "# BLAST - RNF130\n",
    "from Bio import SeqIO\n",
    "from Bio.Blast import NCBIWWW\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import SearchIO\n",
    "\n",
    "\n",
    "def get_prot(id):\n",
    "    \"\"\"\n",
    "    id: Código da proteína a pesquisar na base de dados Swiss-Prot (\"swiss\")\n",
    "    \n",
    "    Retorna uma lista de Accession numbers filtrados em gb e sem repetições\n",
    "    \"\"\"\n",
    "    from Bio import ExPASy\n",
    "    with ExPASy.get_sprot_raw(id) as handle:\n",
    "        seq_record = SeqIO.read(handle, \"swiss\")\n",
    "    print(seq_record.seq)\n",
    "    print(len(seq_record.seq), 'aa')\n",
    "    for k, v in seq_record.annotations.items():\n",
    "        print(k, v)\n",
    "    tam= len(seq_record.seq)\n",
    "    seq= seq_record.seq\n",
    "    tax= seq_record.annotations[\"taxonomy\"]\n",
    "    org= seq_record.annotations[\"organism\"]\n",
    "    y = ('ID:' + id + '|' + 'SEQUENCE:' + seq + '|' + 'SEQUENCE LENGTH:' + str(tam) + 'bp' + '|' + 'TAXONOMY:' + str(tax) + '|' + 'ORGANISM:' + org )\n",
    "    return y\n",
    "\n",
    "def filtro(seq):\n",
    "    \"\"\" \n",
    "    seq: string resultante do get_prot(id)\n",
    "    \n",
    "    Retorna uma lista da string separada pelo caractere \"|\"\n",
    "    \"\"\"\n",
    "    seq = seq.split('|') # Separador \n",
    "    return seq\n",
    "\n",
    "def blast_prot(file, seq):\n",
    "    \"\"\"\n",
    "    file: Nome do ficheiro a dar com o resultado do BLAST\n",
    "    seq: sequência query para executar o BLAST\n",
    "    \n",
    "    Retorna uma lista dos resultados do BLAST com o nome do alinhamento, comprimento do alinhamento e e-value \n",
    "    (filtrado pelo E-Value)\n",
    "    \"\"\"\n",
    "    result_handle = NCBIWWW.qblast('blastp', 'nr', seq)\n",
    "    with open(file, \"w\") as out_handle:\n",
    "        out_handle.write(result_handle.read())\n",
    "    result_handle.close()\n",
    "\n",
    "def obter_homologos(file, E_VALUE_THRESH):\n",
    "    \"\"\"\n",
    "    file: Nome do ficheiro a dar com o resultado do BLAST\n",
    "    E_VALUE_THRESH: Recebe um valor \"None\" ou qualquer número int, corresponde ao e-value máximo aceitável para \n",
    "    tratamento do output do BLAST\n",
    "    \n",
    "    Retorna uma lista dos resultados do BLAST com o nome do alinhamento, comprimento do alinhamento e e-value \n",
    "    (filtrado pelo E-Value)\n",
    "    \"\"\"\n",
    "    result_handle = open(file)\n",
    "    blast_record = NCBIXML.read(result_handle)\n",
    "    FILE = str('SeqsHomologas_' + str(file) + '.fasta')\n",
    "    ficheiro_output = open(FILE, 'w+')\n",
    "    if E_VALUE_THRESH == None:\n",
    "        E_VALUE_THRESH = 0.05\n",
    "    for alignment in blast_record.alignments:\n",
    "        for hsp in range(len(alignment.hsps)):\n",
    "            if alignment.hsps[hsp].expect < E_VALUE_THRESH:\n",
    "                if hsp != 0:\n",
    "                    pass\n",
    "                else:\n",
    "                    ficheiro_output.write('>' + alignment.title + '\\n' + alignment.hsps[hsp].sbjct + '\\n')\n",
    "    result_handle.close()\n",
    "\n",
    "def parse_prot(file):\n",
    "    \"\"\"\n",
    "    file: Nome do ficheiro a dar com o resultado do BLAST\n",
    "    \n",
    "    Retorna informações contidas no ficheiro\n",
    "    \"\"\"\n",
    "    result_handle = open(file)\n",
    "    from Bio.Blast import NCBIXML\n",
    "    blast_record = NCBIXML.read(result_handle)\n",
    "    E_VALUE_THRESH = 0.04\n",
    "    for alignment in blast_record.alignments:\n",
    "        for hsp in alignment.hsps:\n",
    "            if hsp.expect < E_VALUE_THRESH:\n",
    "                print(\"****Alignment****\")\n",
    "                print(\"sequence:\", alignment.title)\n",
    "                print(\"length:\", alignment.length)\n",
    "                print(\"e value:\", hsp.expect)\n",
    "                print(hsp.query[0:75] + \"...\")\n",
    "                print(hsp.match[0:75] + \"...\")\n",
    "                print(hsp.sbjct[0:75] + \"...\")\n",
    "    from Bio import SearchIO\n",
    "    blast_qresult = SearchIO.read(file, \"blast-xml\")\n",
    "    print(blast_qresult)\n",
    "    result_handle.close()\n",
    "\n",
    "\n",
    "def proteico(id, file, blast = False, E_VALUE_THRESH = None):\n",
    "    \"\"\"\n",
    "    genbank: ficheiro genbank (formato .gb)\n",
    "    id: id Swiss-Prot da proteína\n",
    "    file: Nome do ficheiro a dar com o resultado do BLAST\n",
    "    blast: Em caso de omissão, blast = False; se blast = True, irá recorrer ao ficheiro .gb para realizar o BLAST\n",
    "    E_VALUE_THRESH: Recebe um valor \"None\" ou qualquer número int, corresponde ao e-value máximo aceitável para \n",
    "    tratamento do output do BLAST\n",
    "\n",
    "    Gera um ficheiro .txt contendo os Accession Numbers dos resultados do BLAST com filtração segundo o E_VALUE_THRESH e gb\n",
    "    \"\"\"\n",
    "    if blast == True:\n",
    "        from Bio import ExPASy\n",
    "        with ExPASy.get_sprot_raw(id) as handle:\n",
    "            seq_record = SeqIO.read(handle, \"swiss\")\n",
    "        blast_prot(file, seq_record.seq)\n",
    "    x = get_prot(id)\n",
    "    print(x)\n",
    "    obter_homologos(file, E_VALUE_THRESH)\n",
    "    parse_prot(file)\n",
    "\n",
    "proteico('Q86XS8', 'RNF130_blast.xml', True, 0.05)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
